src/process/posix/SDL_posixprocess.c

```diff
@@ -0,0 +1,332 @@
+/*
+  Simple DirectMedia Layer
+  Copyright (C) 1997-2024 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+
+#ifdef SDL_PROCESS_POSIX
+
+#include <fcntl.h>
+#include <errno.h>
+#include <signal.h>
+#include <spawn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/wait.h>
+
+#include "../SDL_sysprocess.h"
+#include "../../file/SDL_iostream_c.h"
+
+#if defined(SDL_PLATFORM_MACOS)
+#include <crt_externs.h>
+#define environ (*_NSGetEnviron())
+#elif defined(SDL_PLATFORM_FREEBSD)
+#include <dlfcn.h>
+#define environ ((char **)dlsym(RTLD_DEFAULT, "environ"))
+#else
+extern char **environ;
+#endif
+
+#define READ_END 0
+#define WRITE_END 1
+
+struct SDL_ProcessData {
+    pid_t pid;
+};
+
+static void CleanupStream(void *userdata, void *value)
+{
+    SDL_Process *process = (SDL_Process *)value;
+    const char *property = (const char *)userdata;
+
+    SDL_ClearProperty(process->props, property);
+}
+
+static bool SetupStream(SDL_Process *process, int fd, const char *mode, const char *property)
+{
+    // Set the file descriptor to non-blocking mode
+    fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK);
+
+    FILE *fp = fdopen(fd, mode);
+    if (!fp) {
+        return false;
+    }
+
+    SDL_IOStream *io = SDL_IOFromFP(fp, true);
+    if (!io) {
+        return false;
+    }
+
+    SDL_SetPointerPropertyWithCleanup(SDL_GetIOProperties(io), "SDL.internal.process", process, CleanupStream, (void *)property);
+    SDL_SetPointerProperty(process->props, property, io);
+    return true;
+}
+
+static bool CreatePipe(int fds[2])
+{
+    if (pipe(fds) < 0) {
+        return false;
+    }
+
+    // Make sure the pipe isn't accidentally inherited by another thread creating a process
+    fcntl(fds[0], F_SETFD, fcntl(fd, F_GETFD) | FD_CLOEXEC);
```

Comment:
Each end is independent so you need something like:

```suggestion
    fcntl(fds[0], F_SETFD, fcntl(fds[0], F_GETFD) | FD_CLOEXEC);
    fcntl(fds[1], F_SETFD, fcntl(fds[1], F_GETFD) | FD_CLOEXEC);
```

This is assuming that `fcntl F_GETFD` can never fail, which is ... probably OK in practice?

Reply:
What smcv said, but let's use the {READ,WRITE}_END macros.
```suggestion
    fcntl(fds[READ_END], F_SETFD, fcntl(fds[READ_END], F_GETFD) | FD_CLOEXEC);
    fcntl(fds[WRITE_END], F_SETFD, fcntl(fds[WRITE_END], F_GETFD) | FD_CLOEXEC);
```